panel/models/jstree.ts

```diff
@@ -0,0 +1,311 @@
+  [key: string]: any
+}
+
+declare function jQuery(...args: any[]): any
+
+export class NodeEvent extends ModelEvent {
+  constructor(readonly data: any) {
+    super()
+  }
+
+  protected override get event_values(): Attrs {
+    return {model: this.origin, data: this.data}
+  }
+
+  static {
+    this.prototype.event_name = "node_event"
+  }
+}
+
+function sync_state(node: any, tree: any) {
+  const node_json = tree.get_node(node.id)
+  if (node_json) {
+    node.state = Object.assign({}, node_json.state, (node.state || {}))
+    node.state.loading = false
+  } else if (!node.state) {
+    node.state = {loading: false}
+  }
+  if (node.children && node.children.length) {
+    for (const child of (node.children || [])) {
+      sync_state(child, tree)
+    }
+  }
+}
+
+export class jsTreeView extends LayoutDOMView {
+  declare model: jsTree
+  protected _container: HTMLDivElement
+  protected _id: any
+  protected _jstree: any
+  protected _setting: boolean
+
+  override connect_signals(): void {
+    super.connect_signals()
+    const {nodes, checked, checkbox, show_icons, show_dots, multiple} = this.model.properties
+    this.on_change(nodes, () => {
+      const tree = this._jstree.jstree(true)
+      for (const node of this.model.nodes) {
+        sync_state(node, tree)
+      }
+      tree.load_node("#", () => {
+        tree.refresh({
+          skip_loading: false,
+          forget_state: false,
+        })
+        this._update_selection_from_value()
+      })
+    })
+    this.on_change(checked, () => {
+      if (!this._setting) {
+        this._update_selection_from_value()
+      }
+    })
+    this.on_change(checkbox, () => this.setCheckboxes())
+    this.on_change(show_icons, () => this._setShowIcons())
+    this.on_change(show_dots, () => this._setShowDots())
+    this.on_change(multiple, () => this._setMultiple())
+  }
+
+  get child_models(): LayoutDOM[] {
+    return []
+  }
+
+  get plugins(): string[] {
+    const plugins = [...this.model.plugins]
+    if (this.model.checkbox && !this.model.plugins.includes("checkbox")) {
+      plugins.push("checkbox")
+    }
+    if (this.model.checkbox && !this.model.plugins.includes("sort")) {
+      plugins.push("sort")
+    }
+    if (!this.model.plugins.includes("wholerow")) {
+      plugins.push("wholerow")
+    }
+    return plugins
+  }
+
+  override stylesheets(): StyleSheetLike[] {
+    return [...super.stylesheets(), jstree_css]
+  }
+
+  override render(): void {
+    super.render()
+    this._id = ID()
+    this._container = div({id: this._id, style: "overflow: auto; minHeight: 200px; minWidth: 200px;"})
+    this.shadow_el.appendChild(this._container)
+    this._jstree = jQuery(this._container).jstree({
+      allow_reselect: true,
+      checkbox: {
+        three_state: this.model.cascade,
+        cascade: this.model.cascade ? "undetermined" : "down+undetermined",
+        cascade_to_disabled: false,
+        tie_selection: false,
+        whole_node: false,
+      },
```

Comment:
I think we need more options for cascade. Especially for generic trees outside of just file trees. I am using my version of jstree right now for something at work to model some genome taxonomy and trying to use the panel-jstree widget I was building until this is ready to be able to cleanly subselect by any level of taxonomy (remove all shown samples with this class, phylum, domain, or species) for example. 

So when deselecting a class, everything below it should deselect, but when selecting a certain domain, it shouldn't select everything below it. That behavior isn't achievable right now with these two cascade options, and I think for generic, non file tree use cases, that would be something wanted.

Reply:
I am rethinking what I want in my own situation actually 🤣 but probably the option to choose more cascade choices would still be good.